src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -176,6 +176,11 @@ QueryPlanPtr MergeTreeDataSelectExecutor::read(
                 "No projection is used when allow_experimental_projection_optimization = 1 and force_optimize_projection = 1",
                 ErrorCodes::PROJECTION_NOT_USED);
 
+        if (settings.force_optimize_projection && settings.aggregate_functions_null_for_empty)
```

Comment:
This part looks useless.

Reply:
Why useless?

I treat it this way: if a user requested "force_optimize_projection" it should know if the optimization is not applied.

Reply:
> if a user requested "force_optimize_projection" it should know if the optimization is not applied.

The exception thrown a few lines above carries such meaning exactly.

If there is some projection define in the table, and for any reason (like `aggregate_functions_null_for_empty = true`) it cannot be used, the exception will be thrown.

The setting was to mimic the behavior of `force_primary_key`, which has the description: `Throw an exception if there is primary key in a table, and it is not used.`. However, I just did a quick test and found that it will throw even when table doesn't have primary key. Not sure whether we should change the description or the behavior.

Reply:
> However, I just did a quick test and found that it will throw even when table doesn't have primary key. Not sure whether we should change the description or the behavior.

Better to change the behavior. (Don't throw if no primary key)